File: local_server.js
const http = require('http');
const { exec } = require('child_process');

const PORT = 3000;

const server = http.createServer((req, res) => {
  // We allow CORS so the extension can fetch this from the background script
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');

  if (req.method === 'OPTIONS') {
    res.writeHead(204);
    res.end();
    return;
  }

  if (req.method === 'POST' && req.url === '/shutdown') {
    console.log('[Local Server] Nhận yêu cầu tắt máy từ Extension!');
    
    // Thực thi lệnh tắt máy trên Windows
    exec('shutdown /s /t 0', (error, stdout, stderr) => {
      if (error) {
        console.error(`[Local Server] Lỗi khi thực hiện tắt máy: ${error.message}`);
        res.writeHead(500, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ success: false, message: 'Lỗi khi tắt máy' }));
        return;
      }
      
      console.log(`[Local Server] Lệnh tắt máy đã được gửi thành công.`);
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ success: true, message: 'Đang tắt máy...' }));
    });
  } else {
    res.writeHead(404, { 'Content-Type': 'text/plain' });
    res.end('Not Found');
  }
});

server.listen(PORT, () => {
  console.log(`[Local Server] Đang chạy tại http://localhost:${PORT}`);
  console.log(`[Local Server] Đợi lệnh tắt máy từ TPlus Auto Extension...`);
});

File: local_server.py
import http.server
import socketserver
import os
import json

PORT = 3000

class ShutdownHandler(http.server.SimpleHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.end_headers()

    def do_POST(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        if self.path == '/shutdown':
            print("[Local Server] Nhận yêu cầu tắt máy từ Extension!")
            
            # Thực thi lệnh tắt máy trên Windows
            status = os.system("shutdown /s /t 0")
            
            if status == 0:
                print("[Local Server] Lệnh tắt máy đã được gửi thành công.")
                self.send_response(200)
                self.send_header('Content-type', 'application/json')
                self.end_headers()
                response = {"success": True, "message": "Đang tắt máy..."}
                self.wfile.write(json.dumps(response).encode('utf-8'))
            else:
                print("[Local Server] Lỗi khi thực hiện tắt máy.")
                self.send_response(500)
                self.send_header('Content-type', 'application/json')
                self.end_headers()
                response = {"success": False, "message": "Lỗi khi tắt máy"}
                self.wfile.write(json.dumps(response).encode('utf-8'))
        else:
            self.send_response(404)
            self.end_headers()
            self.wfile.write(b"Not Found")

    def log_message(self, format, *args):
        # Tắt log mặc định của Python server để màn hình gọn gàng hơn
        pass

with socketserver.TCPServer(("", PORT), ShutdownHandler) as httpd:
    print(f"[Local Server] Đang chạy tại http://localhost:{PORT}")
    print("[Local Server] Đợi lệnh tắt máy từ TPlus Auto Extension...")
    httpd.serve_forever()
